### DISCUSSION TITLE
Dependency Injection for base classes

### DISCUSSION THREAD

**Discussion by suspectinside** [fastapi/fastapi#8625] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# JUST SIMPLE EXAMPLE WITH AFFECTED CLASSES - NO ANY DOMAIN LOGIC PROVIDED

async def connection_factory():
	return 'connection object'

class DBService:
	def __init__(self, connection_factory = Depends(connection_factory)):
		self.conn_factory = connection_factory
	
	async def get_connection(self):
		return self.conn_factory()

class SimpleServiceOne:
	def __init__(self):
		print('SimpleServiceOne.__init__() called')

class SimpleServiceTwo:
	def __init__(self):
		print('SimpleServiceTwo.__init__() called')

class BaseUserService:
	def __init__(self, db_service: DBService = Depends()) -> None:
		self.db_service = db_service

	async def checkAuth(self, id: int):
		return f'BaseUserService.checkAuth() called'

	async def validateCache(self, id: int):
		return f'BaseUserService.validateCache() called'

class UserService(BaseUserService):
	def __init__(self,
		simple_dep_one: SimpleServiceOne = Depends(),
		simple_dep_two: SimpleServiceTwo = Depends(),
	) -> None:
		super().__init__()

		# this would print: Depends(NoneType)
		# but i expect DBService instance injected for __init__ from the base class BaseUserService
		print(f'SETTLED FROM BASE: self.db_service: {self.db_service}')

		self.simple_dep_one = simple_dep_one
		self.simple_dep_two = simple_dep_two
```

### Description

Is it possible to provide (inject) dependency into base class using Depends()?

in the code example in the constructor i receive deps for **simple_dep_one** & **simple_dep_two** but **db_service** contains **self.db_service: Depends(NoneType)**

Its possible to use alternative approach, smtn like:

```python
class UserService(BaseUserService):
	def __init__(self,
		db_service: DBService = Depends(),	# base class dep
		simple_dep_one: SimpleServiceOne = Depends(),
		simple_dep_two: SimpleServiceTwo = Depends(),
	) -> None:
		super().__init__(db_service)

```
but this is very ugly and fastapi's IoC Container should manage all the deps for all the inheritance hierarchy by himself

so, what's the correct way to inject all required dependencies for all possible base classes using fastapi's **Depends()** ?

### Operating System

Linux

### FastAPI Version

0.75.0

### Python Version

3.10

**Comment by Euraxluo** [comment]
you can use injectable

**Comment by suspectinside** [comment]
> you can use injectable

well, if we gonna talk about external libraries, obviously i'll choose [Dependency Injector](https://github.com/ets-labs/python-dependency-injector) (which already has support for FastAPI too), but FastAPI by himself has basic IoC capabilities and IMO it would be much better not to use any external deps but advance FastAPI's "native" features... this is very basic and simple mod why do i have to use external deps

**Reply by nightblure** [reply]
[Dependency Injector](https://github.com/ets-labs/python-dependency-injector) is better than fastapi depends

**Comment by Ar0010r** [comment]
so there is still no other way,  but to use external libraries, right ?

**Comment by dolfinus** [comment]
How UserService is used in this example? Do you use `user_service: UserService = Depends()` in some FastAPI endpoint, or some part of code which is not related to FastAPI at all?

**Comment by lluisotavio** [comment]
Some advance on this topic?

**Comment by YuriiMotov** [comment, marked as answer]
Common approach is to create dependency functions that use sub-dependencies:
```python
async def get_db_service_dep(conn_factory = Depends(connection_factory)):
	return DBService(connection_factory=conn_factory)


async def get_user_service_dep(
    simple_dep_one: SimpleServiceOne = Depends(),
    simple_dep_two: SimpleServiceTwo = Depends(),
    db_service: DBService = Depends(get_db_service_dep),
):
	return UserService(
		simple_dep_one=simple_dep_one,
		simple_dep_two=simple_dep_two,
		db_service=db_service,
    )
```

Full code in details:

<details>

```python
from fastapi import Depends


async def connection_factory():
	return 'connection object'

class DBService:
	def __init__(self, connection_factory):
		self.conn_factory = connection_factory

	async def get_connection(self):
		return self.conn_factory()

class SimpleServiceOne:
	def __init__(self):
		print('SimpleServiceOne.__init__() called')

class SimpleServiceTwo:
	def __init__(self):
		print('SimpleServiceTwo.__init__() called')

class BaseUserService:
	def __init__(self, db_service: DBService) -> None:
		self.db_service = db_service

	async def checkAuth(self, id: int):
		return 'BaseUserService.checkAuth() called'

	async def validateCache(self, id: int):
		return 'BaseUserService.validateCache() called'

class UserService(BaseUserService):
	def __init__(self,
		simple_dep_one: SimpleServiceOne,
		simple_dep_two: SimpleServiceTwo,
		db_service: DBService,
	) -> None:
		super().__init__(db_service=db_service)

		# this would print: Depends(NoneType)
		# but i expect DBService instance injected for __init__ from the base class BaseUserService
		print(f'SETTLED FROM BASE: self.db_service: {self.db_service}')

		self.simple_dep_one = simple_dep_one
		self.simple_dep_two = simple_dep_two


async def get_db_service_dep(conn_factory = Depends(connection_factory)):
	return DBService(connection_factory=conn_factory)


async def get_user_service_dep(
    simple_dep_one: SimpleServiceOne = Depends(),
    simple_dep_two: SimpleServiceTwo = Depends(),
    db_service: DBService = Depends(get_db_service_dep),
):
	return UserService(
		simple_dep_one=simple_dep_one,
		simple_dep_two=simple_dep_two,
		db_service=db_service,
    )
```

</details>

**Reply by purepani** [reply]
The details didn't seem to format correctly, if you could fix that 🙂
